src/server/scripts/Pet/pet_dk.cpp

```diff
@@ -139,7 +157,263 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
+                    default:
+                        break;
+                }
+            }
+
+            void UpdateAI(uint32 diff) override
+            {
+                if (me->HasReactState(REACT_PASSIVE))
+                {
+                    if (_engageTimer <= diff)
+                        me->SetReactState(REACT_AGGRESSIVE);
+                    else
+                    {
+                        _engageTimer -= diff;
+                        return;
+                    }
+                }
+
+                if (me->IsInCombat() && (!me->GetVictim() || !me->IsValidAttackTarget(me->GetVictim())))
+                    EnterEvadeMode(EVADE_REASON_NO_HOSTILES);
+
+                if (!me->IsInCombat())
+                {
+                    Unit* ownerTarget = nullptr;
+                    if (Player* owner = me->GetCharmerOrOwner()->ToPlayer())
+                        ownerTarget = owner->GetSelectedUnit();
+
+                    // recognize which victim will be choosen
+                    if (ownerTarget && ownerTarget->GetTypeId() == TYPEID_PLAYER)
+                    {
+                        if (!ownerTarget->HasBreakableByDamageCrowdControlAura(ownerTarget))
+                            Attack(ownerTarget);
+                    }
+                    else if (ownerTarget && (ownerTarget->GetTypeId() != TYPEID_PLAYER) && IsInThreatList(ownerTarget))
+                    {
+                        if (!ownerTarget->HasBreakableByDamageCrowdControlAura(ownerTarget))
+                            Attack(ownerTarget);
+                    }
+                    else
+                        Init();
+                }
+
+                /*
+                    Investigate further if these casts are done by
+                    any owned aura, eitherway SMSG_SPELL_GO
+                    is sent every X seconds.
+                */
+                _events.Update(diff);
+
+                while (uint32 _eventId = _events.ExecuteEvent())
+                {
+                    switch (_eventId)
+                    {
+                        case EVENT_SPELL_CAST_1:
+                            // Cast every second
+                            if (Unit* victim = me->GetVictim())
+                                DoCast(victim, SPELL_AGGRO_8_YD_PBAE, true);
+                            _events.ScheduleEvent(EVENT_SPELL_CAST_1, 1 * IN_MILLISECONDS);
+                            break;
+                        case EVENT_SPELL_CAST_2:
+                            // Cast every 6 seconds
+                            DoCast(me, SPELL_DK_DANCING_RUNE_WEAPON_VISUAL, true);
+                            _events.ScheduleEvent(EVENT_SPELL_CAST_2, 6 * IN_MILLISECONDS);
+                            break;
+                        default:
+                            break;
+                    }
+                }
+
+                if (me->IsInCombat() && me->GetVictim() && me->IsValidAttackTarget(me->GetVictim()))
```

Comment:
I wonder why Unit::IsValidAttackTarget() doesn't NULL-checks on its own to avoid thousands of NULL checks all around the code